### PR TITLE
FEAT-#1200: pivot_table implementation

### LINKED ISSUE #1200: pivot_table function implementation

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1200  
- [x] tests added and passing

### CODE REVIEW

**Comment by dchigarev** [comment]
@devin-petersohn I've faced with some problems trying to group by on columns, that have `NaN` values in my `pivot_table` implementation. The problem is that [`get_indices`](https://github.com/modin-project/modin/blob/master/modin/engines/base/frame/partition_manager.py#L320-L349) seems to be losing some index meta-information that further required for proper `unstack` result (`get_indices` called by [`_apply_full_axis`](https://github.com/modin-project/modin/blob/master/modin/engines/base/frame/data.py#L935-L979) which called by [`groupby_agg`](https://github.com/modin-project/modin/blob/master/modin/backends/pandas/query_compiler.py#L1546-L1584)). Look at that reproduce:
```python
if __name__ == "__main__":
    import modin.pandas as pd
    import pandas
    import numpy as np

    pd.DEFAULT_NPARTITIONS = 4

    data1 = {
        "NaN_col1": [None] + np.arange(255).tolist(),
        "NaN_col2": np.arange(128).tolist() + [None] + np.arange(127).tolist(),
        "Data_Col": np.arange(256),
    }

    df = pd.DataFrame(data1)

    mask = [df["NaN_col1"]._to_pandas(), df["NaN_col2"]._to_pandas()]

    qc = df._query_compiler.groupby_agg(
        by=mask,
        axis=0,
        agg_func=lambda df: df.agg(np.mean),
        groupby_args={},
        agg_args={},
    )                                                                 # Output:
    pandas_res = df._to_pandas().groupby(by=mask).agg(np.mean)        #
								      #
    print("md.index == pd.index", all(qc.index == pandas_res.index))  # md.index == pd.index True
								      #
    print("md_len:", len(qc.index.levels[0]))                         # md_len: 254
    print("pd_len:", len(pandas_res.index.levels[0]))		      # pd_len: 255
								      #
    print("127 in md.levels", 127 in qc.index.levels[0])	      # 127 in md.levels False
    print("127 in pd.levels", 127 in pandas_res.index.levels[0])      # 127 in pd.levels True

```
Indices itself seems to be equals, but the levels does not.

There is nothing special about `groupby_agg`, it works correctly, if we'll set breakpoint [here](https://github.com/modin-project/modin/blob/master/modin/engines/ray/pandas_on_ray/frame/partition_manager.py#L86) (in case of `Ray` execution engine) we'll be able to see a place where information about levels loses:
```
-> return new_idx[0].append(new_idx[1:]) if len(new_idx) else new_idx
(Pdb) len(new_idx[0].levels[0])				# levels of original indices seems correct
255
(Pdb) len(new_idx[0].append(new_idx[1:]).levels[0])     # but after appending, information about levels are missing
254

```
It looks like `append` is not working well in that case. I've briefly searched through Pandas Index documentation and didn't find any alternatives of `append`. There is [`join`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Index.join.html) and [`union`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Index.union.html) but both of them didn't give desired result. 

For now, I've let `pivot_table` defaulting to pandas if any of columns to group by has `NaN` values.
@devin-petersohn Any ideas how to keep that meta information? Or should we leave it default to pandas in that case?

**Comment by dchigarev** [comment]
The issue described in [that comment](https://github.com/modin-project/modin/pull/1669#issuecomment-655018358) was fixed by creating the correct index manually in case of `NaN` groups. I guess that now the full functionality of `pivot_table` was implemented in ac5105c. 

Note that this implementation based on using `unstack` function that will be implemented in #1649 (for now it's `default_to_pandas`). 

Also, that branch requires fixes from #1727, so that branch also contains from #1727 to make CI green, after #1727 will be merged that branch should be rebased.

Inserting multiple columns (margins in the pivoted frame) were implemented with [`_sorted_multi_insert`](https://github.com/modin-project/modin/pull/1669/commits/47e15c2c6eccccb3715be88e80d5412dda55f75e#diff-db8dc531e116caee2b9cc5ccbcc5c642R1890-R1980) assuming that fast `sort_index` will be implemented soon, it also could be optimized after merging #1773, that will allow to insert distributed columns in a parallel way.

For now @devin-petersohn I need your review

**Comment by anmyachev** [comment]
@dchigarev any progress on this?

**Comment by dchigarev** [comment]
> @dchigarev any progress on this?

Not yet, this issue is in backlog for now

**Comment by dchigarev** [comment]
@devin-petersohn this PR is ready for review. Here is a note about what this PR is adding:

Essentially `pivot_table` is a groupby aggregation + unstack. Since we need to do a groupby aggregation we need knowledge about the full index axis. And since we're gonna groupby on `index + columns` columns we must broadcast these columns to every column partition. So what this PR is doing is adding the ability to broadcast multi-partitioned axis partitions and using this to apply `pandas.DataFrame.pivot_table` with broadcasting `index + columns` columns to the columns partitions of the source frame.

Here is the results of time measurements:
```
200,000x256
Modin: 2.33s
Pandas: 1.37s

500,000x256
Modin: 2.75s
Pandas: 2.46s

1,000,000x256
Modin: 3.46s
Pandas: 4.18s
```

**Comment by dchigarev** [comment]
@devin-petersohn just a reminder to review

**Comment by dchigarev** [comment]
@devin-petersohn @YarShev just a reminder to review)

**Comment by dchigarev** [comment]
in continuous of [this](https://github.com/modin-project/modin/pull/1669#issuecomment-691692943) message:

@YarShev noticed, that columns is not the only axis, on which the result of `pivot_table` could be different from pandas. In some cases, pandas transposes the result of `pivot_tabel`, so columns (unsorted in case of modin) became indices (which still be unsorted in modin).

So this is how current `pivot_table` implementation in Modin differs from Pandas:

1. If Modin `pivot_tables` obtains parameters with that properties: `index is None and len(columns) > 1` then the resulted frame will have unsorted indices (when pandas produces frame with sorted one)
2. With all other parameters, Modin `pivot_table` produces frame with unsorted columns (when pandas produces frame with sorted one)

We're getting these mismatches because of the limitations of applying `pivot_table`, described in the [previous comment](https://github.com/modin-project/modin/pull/1669#issuecomment-691692943).

So we should decide what's more important for us, time that we saved when we are not sorting one of the axis, or sorted columns at the `pivot_table` result.

@devin-petersohn @YarShev what are your thoughts?

File: modin/backends/pandas/query_compiler.py
Comment (YarShev) [inline modin/backends/pandas/query_compiler.py:2594]: Isn't the one transposing enough? As I see, we transpose result in this line and in `applyier` as well.

File: modin/engines/base/frame/data.py
Comment (YarShev) [inline modin/engines/base/frame/data.py:966]: Why are `args` and `kwargs` needed?

File: modin/engines/base/frame/data.py
Comment (YarShev) [inline modin/engines/base/frame/data.py:1554]:
```suggestion
        Parameters
        ----------
```

File: modin/engines/base/frame/data.py
Comment (YarShev) [inline modin/engines/base/frame/data.py:1570]:
```suggestion
        Returns
        -------
```

File: modin/engines/base/frame/partition_manager.py
Comment (YarShev) [inline modin/engines/base/frame/partition_manager.py:218]: This doesn't correspond numpydoc

File: modin/engines/base/frame/axis_partition.py
Comment (YarShev) [inline modin/engines/base/frame/axis_partition.py:151]: Why is `reversed` needed?

File: modin/engines/base/frame/axis_partition.py
Comment (YarShev) [inline modin/engines/base/frame/axis_partition.py:271]: Can you describe a little bit this logic and what is this needed for?

File: modin/engines/base/frame/data.py
Comment (YarShev) [reply modin/engines/base/frame/data.py:1554]: Description of `other` is missed.

File: modin/backends/pandas/query_compiler.py
Comment (YarShev) [inline modin/backends/pandas/query_compiler.py:2587]: Can we convert `keys_columns` to pandas and apply `applyier` on df and other?

Comment (YarShev) [reply modin/backends/pandas/query_compiler.py:2587]: I mean the approach which is used in `merge` implementation for example.

**Comment by YarShev** [comment]
> in continuous of [this](https://github.com/modin-project/modin/pull/1669#issuecomment-691692943) message:
> 
> @YarShev noticed, that columns is not the only axis, on which the result of `pivot_table` could be different from pandas. In some cases, pandas transposes the result of `pivot_tabel`, so columns (unsorted in case of modin) became indices (which still be unsorted in modin).
> 
> So this is how current `pivot_table` implementation in Modin differs from Pandas:
> 
> 1. If Modin `pivot_tables` obtains parameters with that properties: `index is None and len(columns) > 1` then the resulted frame will have unsorted indices (when pandas produces frame with sorted one)
> 2. With all other parameters, Modin `pivot_table` produces frame with unsorted columns (when pandas produces frame with sorted one)
> 
> We're getting these mismatches because of the limitations of applying `pivot_table`, described in the [previous comment](https://github.com/modin-project/modin/pull/1669#issuecomment-691692943).
> 
> So we should decide what's more important for us, time that we saved when we are not sorting one of the axis, or sorted columns at the `pivot_table` result.
> 
> @devin-petersohn @YarShev what are your thoughts?

If we want to have a parallel implementation of `pivot_table` for the upcoming release, we can merge this implementation after review done. However, I think we do many assumptions related to incorrect order of rows and columns. That's why this implementation should be prioritized as P1 in order to fix this for future releases. @devin-petersohn , what do you think?

**Comment by YarShev** [comment]
One more question. @dchigarev , what problems are occurred if we do row partitions instead of column partitions?

**Comment by dchigarev** [comment]
> One more question. @dchigarev , what problems are occurred if we do row partitions instead of column partitions?

`pivot_table` aggregates columns data, so under the hood it's doing something like this: 
```python
df.groupby(by=[index]+[columns]).aggregate(aggfunc).unstack(level=[columns])
```
which requires us to have the whole column at the same place to aggregate it. We also can't apply an aggregation function to every row partition and then for their results, because we can obtain a function, that is not [associative](https://en.wikipedia.org/wiki/Associative_property) (for example `mean`), which means that `a*b*c*d != (a*b)*(a*c)`

File: modin/backends/pandas/query_compiler.py
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:2587]: I'm not sure, but I think for some engines it could be slower to do `.to_pandas` for the frame rather than broadcasting it. I'm realizing that there is no difference in the shape of the broadcasted frame, which still be N full columns. 

The benefit in broadcasting here, is that we're passing to the function not serialized objects, [but its `oid`](https://github.com/dchigarev/modin/blob/1f8ee522999f7a3773e0b42ecc8b3cd56ef54137/modin/engines/ray/pandas_on_ray/frame/axis_partition.py#L22-L27) (in case of `Ray`, other engines may have their own names for that). So it's engines responsobility to serialize objects in the most efficient way, when their executing a function.

I don't have any performance measurements on it, and not closely familiar with how `Ray` and other engines obtains futures/objects ids, is this beneficial optimization or not. Let's ask @devin-petersohn since he's more competent in that question

File: modin/backends/pandas/query_compiler.py
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:2594]: We transpose result in `applier` to produce consistent partition. 

All partitions in a row must have the same indices and all partitions in a column must have the same columns. When we're applying our function to a columns partitions we're guarantee that partitions will be consistent. However with that conditions:
```python
if len(index) == 0 and len(columns) > 0
```
pandas transposes result of the `pivot_table` and frame loses it consistency:
```
before pandas transposing (consistent frame):
   a b |   c d
1  1 2 |1  7 8
2  3 4 |2  9 0
3  5 6 |3  1 2

after pandas transposing (inconsistent frame, because of different indices at the same row):
   1 2 3 |   1 2 3
a  1 3 5 |c  7 9 1
b  2 4 6 |d  8 0 2
```
so in `applyier` we transposing the result back to consistent state, and after that we're transposing the whole frame correctly in the query compiler

File: modin/engines/base/frame/data.py
Comment (dchigarev) [reply modin/engines/base/frame/data.py:966]: to add the ability to function to obtain any arguments, like `other`

**Comment by devin-petersohn** [comment]
>If we want to have a parallel implementation of pivot_table for the upcoming release, we can merge this implementation after review done. However, I think we do many assumptions related to incorrect order of rows and columns. That's why this implementation should be prioritized as P1 in order to fix this for future releases. @devin-petersohn , what do you think?

Yes, I agree with this.

**Review by devin-petersohn — COMMENTED** [review]
I agree with @YarShev that we can merge this after the comments, and raise a new issue for next release. We should probably warn the user about the order of the columns mismatching and say that it will be fixed in a future release.

File: modin/engines/base/frame/data.py
Comment (dchigarev) [reply modin/engines/base/frame/data.py]: fixed

File: modin/engines/base/frame/data.py
Comment (dchigarev) [reply modin/engines/base/frame/data.py]: fixed

File: modin/engines/base/frame/partition_manager.py
Comment (dchigarev) [reply modin/engines/base/frame/partition_manager.py]: fixed

**Review by devin-petersohn — COMMENTED** [review]
I am okay with the changes as-is. @YarShev feel free to merge if you have no other comments.

File: modin/backends/pandas/query_compiler.py
Comment (YarShev) [reply modin/backends/pandas/query_compiler.py:2587]: It would be interesting to look at performance measurement these approaches. @dchigarev , can you do some measurements?

Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:2587]: Here are the results of measurements for 200,000 rows DataFrame. It seems that broadcasting is significantly faster rather than `to_pandas()`
```
MEASURING single-partition columns; df shape is (200000, 255)
to_pandas 2.890704656019807
broadcasting 0.8330112155526876


MEASURING multi-partition columns; df shape is (200000, 253)
to_pandas 4.838804990053177
broadcasting 1.0249458067119122
```

<details><summary>script</summary>


```python
import numpy as np
import os
import pandas
from timeit import default_timer as timer


def generate_data_file(
    filename=None,
    nrows=200000,
    ncols=256,
    multiline_rows=False,
    force=False,
    rand_low=-100,
    rang_high=100,
    log=True,
):
    if filename is None:
        filename = os.path.abspath(
            f"int_dataset-{nrows},{ncols},{rand_low},{rang_high},{multiline_rows}.csv"
        )

    print("generating", filename)

    if os.path.exists(filename) and not force:
        return filename

    data = {
        f"col{i}": np.concatenate(
            [
                np.concatenate(
                    [
                        ["some\nvery very very\nlong string\nwith many multilines"],
                        np.random.randint(rand_low, rang_high, 9),
                    ]
                )
                for _ in np.arange(ncols // 10)
            ]
        )
        if (i % 10 == 0 and multiline_rows)
        else list(np.random.randint(rand_low, rang_high, nrows // 10)) * 10
        for i in np.arange(ncols)
    }
    print("dict generated!")
    df = pandas.DataFrame(data)
    print("dataframe created!")
    df.to_csv(filename)
    print("csv ready!")
    return filename


def measure_time(md_df, columns, test_label):
    subframe = md_df[columns]
    md_df = md_df.drop(columns=columns)

    print(f"\n\nMEASURING {test_label}; df shape is {md_df.shape}")

    t1 = timer()
    pd_subframe = subframe._to_pandas()

    def single(df):
        pd_subframe.max()
        df.max()
        return df

    result = pd.DataFrame(
        query_compiler=PandasQueryCompiler(
            md_df._query_compiler._modin_frame._apply_full_axis(0, single)
        )
    )
    repr(result)
    print("to_pandas", timer() - t1)

    t1 = timer()

    def single(df, other):
        other.max()
        df.max()
        return df

    result = pd.DataFrame(
        query_compiler=PandasQueryCompiler(
            md_df._query_compiler._modin_frame.broadcast_apply_full_axis(
                0, single, other=subframe._query_compiler._modin_frame
            )
        )
    )
    repr(result)
    print("broadcasting", timer() - t1)


if __name__ == "__main__":
    import modin.pandas as pd
    from modin.backends import PandasQueryCompiler
    from modin.pandas.test.utils import df_equals, test_data
    import pandas
    import numpy as np

    filename = generate_data_file()

    md_df = pd.read_csv(filename)

    measure_time(md_df, md_df.columns[:2], test_label="single-partition columns")
    measure_time(
        md_df,
        [*md_df.columns[:2], *md_df.columns[-2:]],
        test_label="multi-partition columns",
    )

```



</details>

Comment (YarShev) [reply modin/backends/pandas/query_compiler.py:2587]: @dchigarev , thanks! It's great.

**Comment by YarShev** [comment]
@dchigarev , please, answer rest comments. After that we will merge the PR if there are no objections.

File: modin/engines/base/frame/axis_partition.py
Comment (dchigarev) [reply modin/engines/base/frame/axis_partition.py:151]: We need `other_shape` value to save the shape of broadcasting frame in terms of partitions before flattening it into 1D array. We want to see shape in that order `(nrows, ncols)`. We're computing `other_shape` basing on the length of `other_axis_partition` (array of axis partitions) and on the amount of partitions in axis partition (`list_of_blocks`).

Let's see now, how `other_shape` value depends on axis:

image


image

File: modin/engines/base/frame/axis_partition.py
Comment (dchigarev) [reply modin/engines/base/frame/axis_partition.py]: I've rewrited this part a bit, this is how it's working now:

Here is the order of how partitions of broadcasted frame will be located in `rt_parts` for each axis:
image
as we can see:
```python
other[axis] == len(list_of_blocks)
other[axis ^ 1] == len(other_axis_partitions)  # amount of axis partitions
```
So here we're just building our broadcasted frame from a flattened array.
`combined_axis` will contain an array of serialized axis partitions, and `rt_frame` will contain the whole serialized broadcasted frame

**Comment by YarShev** [comment]
@dchigarev , thanks, LGTM! Please, create an issue to fix the problem related mismatching of rows and columns in the next release.

**Review by YarShev — APPROVED** [review]

**Pull request merged**
